File: miner.pl
#!/usr/bin/env perl
# The following sites don't work:
#  http://en.wikipedia.org/wiki/

use strict;
use warnings;

use Encode::Guess;
use WWW::Mechanize;
use Encode::Guess;
use Encode;
use LWP::Simple;
use HTML::DOM;
use HTML::DOM::Node ':all';
use List::Util 'first';
use Smart::Comments;

sub help {
    die <<_EOC_;
Usage:
  ./miner.pl 'http://so.sohu.com/web?query=Perl' 5
  ./miner.pl 'http://www.baidu.com/s?wd=Perl' 5 gbk
  ./miner.pl 'http://search.yahoo.com/search?p=Perl' 5
  ./miner.pl 'http://www.yahoo.cn/s?p=Perl' 5
  ./miner.pl 'http://www.answers.com/Perl' 5
  ./miner.pl 'http://search.cpan.org/search?query=UML&mode=all' 5
  ./miner.pl 'http://www.amazon.com/s/002-8506712-0260828?ie=UTF8&tag=mozilla-20&index=blended&link%5Fcode=qs&field-keywords=Perl&sourceid=Mozilla-search' 5 utf8
  ./miner.pl 'http://search.ebay.com/Python_W0QQfrppZ50QQfsopZ1QQmaxrecordsreturnedZ300' 5 latin-1
   ./miner.pl 'http://search.atomz.com/search/?sp-q=Perl&sp_a=sp1000a5a9&sp_f=ISO-8859-1&sp_t=general&sp-x-2=v2_cat2&sp-q-2=&sp-c=25&sp-k=&sp-p=all&sp-k=Articles%7CBooks%7CConferences%7COther%7CWeblogs%7CCourses&p=&sp-q-1=&c=&counter=&query=Perl&search=Search' 5
  ./miner.pl 'http://packages.ubuntu.com/cgi-bin/search_packages.pl?searchon=names&subword=1&version=feisty&release=all&keywords=Python&sourceid=mozilla-search' 5


_EOC_
}

my $HitCount = 0;
my $url = shift or help();
my $count = shift || 10;
my $encoding = shift;
my $mech = WWW::Mechanize->new(
    cookie_jar => {},
    autocheck => 1,
);
$mech->agent_alias('Linux Mozilla');
$mech->get($url);
my $html = $mech->content;
if (!defined $encoding) {
    my $decoder = guess_encoding($html, qw/utf8 gbk/);
    die $decoder unless ref($decoder);
    $html = encode('utf8', $decoder->decode($html));
} else {
    $html = encode('utf8', decode($encoding, $html));
}
#$html = decode('UTF8', $html);
my $doc = HTML::DOM->new;
#$doc->utf8_mode(1);
$doc->parse($html);
#print $doc->as_HTML;
my $body = $doc->body;
#print $body->as_HTML;
mine_pattern($body);
warn "\n\ninfo: For total $HitCount hit(s) found.\n";

sub mine_pattern {
    my $node = shift;
    my $prefix = shift || 'body>';
    my $list = $node->childNodes;
    my @categories;
    my $locator;
    my @elems;
    my %sample;
    for my $i (0..$list->length-1) {
        my $elem = $list->item($i);
        next unless $elem->nodeType == ELEMENT_NODE;
        my $tag_name = lc($elem->tagName);
        my ($tag_class) = split /\s+/, $elem->getAttribute("class");
        my $tag_id = $elem->getAttribute("id");
        $locator = $tag_name;

        if ($tag_id) { $locator .= "#$tag_id"; }
        if ($tag_class) { $locator .= ".$tag_class"; }

        my $seq = compute_seq($elem);
        if (!$seq || $seq !~ /\+.*\+.*\+/) {
            next;
        }
        $sample{$seq} .= "[$prefix$locator - $seq] " . $elem->as_text . "\n\n";
        my $category = first { $_->[0] && $_->[0] eq $seq } @categories;
        if (!$category) {
            # create a new category:
            push @categories, [$seq];
        } else {
            # push $seq to the existing category:
            push @$category, $seq;
        }
        push @elems, $elem;
    }
    my @hits = grep {
        #warn scalar(@$_), "\n";
        scalar(@$_) >= $count
    } @categories;

    if (@hits) {
        $HitCount++;
        my $seq = $hits[0]->[0];
        ### @hits
        ### text: $sample{$seq}
        ## @categories
    }
    for my $elem (@elems) {
        mine_pattern($elem, $prefix . $locator . '>');
    }
    return unless @hits;
    #warn $prefix . $locator;
}

sub compute_seq {
    my $node = shift;
    my $html = $node->as_HTML;
    my $s;
    while ($html =~ /<(\/?)\s*(\w+)[^>]*>/gcs or
        $html =~ /<()(\w+)\s*(\/)\s*>/gcs) {
        my ($before, $tag, $after) = ($1, $2);
        $tag = lc($tag);
        next if $tag =~ /^ (?: [biu] |\d+|nobr|wbr|br|span|font|small|big|em|strong|dfn|code|samp|kbd|var|cite|basefont|img|applet|script|noscript|map|area|tt|trike|big|sub|sup ) $/x;
        if ($before) {
            $s .= "-$tag";
        } elsif ($after) {
            $s .= "+$tag-$tag";
        } else {
            $s .= "+$tag";
        }
    }
    return $s;
    #warn ($s =~ s/<(\w+)><\/\1>|<\w+\/>//gs);
    #warn $s;
    #die $node->nodeType;
    #warn "==========================";
    #warn $html;
}

File: js/miner.js
var HitsCount;
var Hostname;

function mine_pattern (doc, count, hostname) {
    var body = $("body", doc)[0];
    Hostname = hostname;
    HitsCount = 0;
    var patterns = mine_node(body, count, 'body>');
    //Debug.JJJ(patterns);
    return patterns;
}

function mine_node (node, count, prefix) {
    //alert(node);
    prefix = prefix || '';
    var list = node.childNodes;
    if (!node) {
        alert(prefix + node);
        return [];
    }
    var categories = [];
    var locator = '';
    var elems = [];
    var samples = {};
    var matched_elems = {};
    for (var i = 0; i < list.length; i++) {
        var elem = list[i];
        if (elem.nodeType != Node.ELEMENT_NODE)
            continue;
        //info(Hostname + ": locator: " + locator);

        var seq = compute_seq(elem);
        //info(Hostname + ": seq: " + seq);

        if (!seq || ! (/\+.*\+.*\+/.test(seq))) {
            //alert(tagName + ": " + seq);
            continue;
        }
        samples[seq] += "[" + seq + "] " + $(elem).text() + "\n\n";
        matched_elems[seq] = elem;
        category = categories.filter(
            function (c) { return c[0] == seq; }
        )[0];
        if (!category) {
            // create a new category:
            categories.push([seq]);
        } else {
            category.push(seq);
        }
        elems.push(elem);
    }
    //alert("elems: " + elems.length);
    //if (Hostname == 'www.yisou.com')
        //info(Hostname + ": Categories:\n" + Dumper(categories));
    var retvals = [];

    var hits = categories.filter(function (e) {
        return e.length >= count;
    });
    if (hits.length > 0) {
        HitsCount++;
        var seq = hits[0][0];
        //info(Hostname + ": hits:\n" + Dumper(hits));
        var pattern = genPattern(matched_elems[seq]);
        retvals.push(pattern);
        info(Hostname + ": pattern:\n" + pattern);
        //info(Hostname + ": sample:\n" + Dumper(samples[seq]));

    } else {
        for (var i = 0; i < elems.length; i++) {
            var elem = elems[i];
            //alert(i + ": " + elem);
            //alert(i + ": " + elem.nodeType);
            var more = retvals.concat(mine_node(elem, count, prefix + locator + ">"));
            //if (more.length > 0)
                //alert("Found more!");
            retvals = retvals.concat(more);
        }
    }
    return retvals;
}

function genPattern (node) {
    var selector = '';
    while (true) {
        if (node == null) break;
        if (node != null) {
            var tagId = node.getAttribute('id');
            var tagName = node.tagName.toLowerCase();
            var tagClass = node.getAttribute('class');
            if (tagClass) {
                tagClass = tagClass.split(/\s+/)[0];
            } else {
                tagClass = "";
            }
            var locator = tagName;
            if (selector != '' && tagId && /^[-\w]+$/.test(tagId))
                locator += "#" + tagId;
            else if (tagClass && /^[-\w]+$/.test(tagClass))
            locator += "." + tagClass;

            if (selector != '')
                selector = locator + ">" + selector;
            else
                selector = locator;
        }
        if (node.tagName == 'BODY') break;
        node = node.parentNode;
    }
    return selector;
}

function compute_seq (node) {
    var html = $(node).html();
    var s = '';
    var tagRe = /<(\/?)\s*(\w+)[^>]*>|<()(\w+)\s*(\/)\s*>/g;
    while ((matches = tagRe.exec(html)) != null) {
        var before = matches[1];
        var tag    = matches[2].toLowerCase();
        var after  = matches[3];
        if (/^(?:[biu]|\d+|nobr|wbr|br|span|font|small|big|em|strong|dfn|code|samp|kbd|var|cite|basefont|img|applet|script|noscript|map|area|tt|trike|big|sub|sup)$/.test(tag))
            continue;
        //info(Hostname + ": tag: " + tag);
        //info(Hostname + ": before: " + before);
        //info(Hostname + ": after: " + after);
        if (before)
            s += "-" + tag;
        else if (after)
            s += "+" + tag + "-" + tag;
        else
            s += "+" + tag;
    }
    return s;
}

